pkg/restore/prioritize_group_version.go

```diff
@@ -0,0 +1,337 @@
+/*
+Copyright The Velero Contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package restore
+
+import (
+	"context"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/version"
+
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/archive"
+	"github.com/vmware-tanzu/velero/pkg/client"
+)
+
+// ChosenGroupVersion is the API Group version that was selected to restore
+// from potentially multiple backed up version enabled by the feature flag
+// APIGroupVersionsFeatureFlag
+type ChosenGroupVersion struct {
+	Group   string
+	Version string
+	Dir     string
+}
+
+// chooseAPIVersionsToRestore will choose a version to restore based on a user-
+// provided config map prioritization or our version prioritization.
+func (ctx *restoreContext) chooseAPIVersionsToRestore() error {
+	sourceGVs, targetGVs, userGVs, err := ctx.gatherSourceTargetUserGroupVersions()
+	if err != nil {
+		return err
+	}
+
+OUTER:
+	for rg, sg := range sourceGVs {
+		// Default to the source preferred version if no other common version
+		// can be found.
+		cgv := ChosenGroupVersion{
+			Group:   sg.Name,
+			Version: sg.PreferredVersion.Version,
+			Dir:     sg.PreferredVersion.Version + velerov1api.PreferredVersionDir,
+		}
+
+		tg := findAPIGroup(targetGVs, sg.Name)
+		if len(tg.Versions) == 0 {
+			ctx.SetChosenGVToRestore(cgv, rg, "", "")
+			continue
+		}
+
+		// Priority 0: User Priority Version
+		if userGVs != nil {
+			uv := findSupportedUserVersion(userGVs[rg].Versions, tg.Versions, sg.Versions)
+			if uv != "" {
+				ctx.SetChosenGVToRestore(cgv, rg, sg.PreferredVersion.Version, uv)
```

Comment:
Thank you for your feedback. I also would not expect a `set` function to be modifying its arguments. And I agree that I should follow that commonly understood convention of `set`. I ended up getting rid of the `set` function and moved the logic into the calling function.

The reason I had the `set` function mutate its arguments is because I thought it would reduce writing repetitive mutation code. However, it came at the cost of readability. 

I believe it's worth repeating some code and logic if it means it's easier to follow. As a result, I have made some changes in accordance to your suggestions.